### ISSUE TITLE
bad compilation error message on macro error

### ISSUE THREAD

**Issue by kostya** [crystal-lang/crystal#3579]
i spend some time to find a bug in big code, error message not informative:
small example:
```crystal
abstract class A
  abstract def bla
end

class B < A
  JSON.mapping({bla: String})
end

p B.from_json(%q<{"bla": "xxx"}>)
```

```
Error in 1.cr:2: abstract `def A#bla()` must be implemented by B

  abstract def bla
               ^~~
```

but actually i forget to `require "json"`

**Comment by puppetpies** [comment]
I got something similar to work as it seems you have to create the abstract def under a module namespace.

require "json"

module Test

  abstract class A
     abstract def bla
     end
  end

  class Test::B < Test::A
    JSON.mapping({bla: String})
  end


p Test::B.from_json(%q<{"bla": "xxx"}>)

**Comment by kostya** [comment]
my example also work if i `require "json"`, this bug about when `abstract def` implementation generated by some macro, which just failed to generate it. then impossible to find why it happen.

**Comment by jhass** [comment]
Yes, the abstract def check shouldn't shadow the missing constant error.
